### ISSUE TITLE
GPU memory estimation before training starts

### ISSUE THREAD

**Issue by NP-compete** [NP-compete/smithery#39]
## Problem

Users don't know if their GPU can handle a recipe until it OOMs 20 minutes into training. By then they've wasted time, electricity, and patience. Smithery should estimate memory requirements upfront and warn before starting.

### Memory Estimator

\`\`\`python
from smithery.train.memory import estimate_vram

estimate = estimate_vram(
    base_model="microsoft/Phi-3.5-mini-instruct",
    method="qlora",
    quantization="4bit",
    lora_r=16,
    batch_size=4,
    max_seq_length=4096,
    gradient_accumulation_steps=4,
)

print(estimate)
# VRAMEstimate(
#   model_weights_gb=2.1,    # 4-bit quantized model
#   lora_adapters_gb=0.3,    # LoRA trainable params
#   optimizer_states_gb=0.6, # AdamW states
#   activations_gb=4.2,      # Batch activations (depends on batch_size * seq_len)
#   overhead_gb=1.0,         # CUDA overhead, fragmentation
#   total_gb=8.2,
#   fits_on=["RTX 3090 (24GB)", "RTX 4090 (24GB)", "A100 (40GB)"],
#   warning=None,
# )
\`\`\`

### Estimation Formula

\`\`\`
Model weights (4-bit):  params * 0.5 bytes
LoRA adapters (fp16):   r * (in_features + out_features) * num_modules * 2 bytes
Optimizer (AdamW):      ~2x LoRA adapter size (momentum + variance)
Activations:            batch_size * seq_len * hidden_size * num_layers * 2 bytes
                        (reduced by gradient checkpointing: ~divide by num_layers)
Overhead:               ~1-2 GB (CUDA context, fragmentation)
\`\`\`

### Pre-training Check

Before training starts, automatically:
1. Estimate required VRAM
2. Check available GPU memory
3. If insufficient: print warning with suggestions
   - Reduce batch_size
   - Reduce max_seq_length
   - Use gradient_accumulation_steps instead of larger batches
   - Use a smaller base model

\`\`\`
⚠️  Estimated VRAM: 18.4 GB | Available: 16.0 GB (RTX 4080)

Suggestions:
  • Reduce batch_size from 4 to 2 (saves ~4 GB)
  • Reduce max_seq_length from 4096 to 2048 (saves ~2 GB)
  • Try configs/phi3-tool-calling.yaml instead (10 GB)

Continue anyway? [y/N]
\`\`\`

### CLI Dry Run

\`\`\`bash
smithery train --config configs/mistral-7b-tools.yaml --data ./data/train.jsonl --dry-run
# Shows: estimated VRAM, training time, cost, data stats — without loading the model
\`\`\`

## Acceptance Criteria

- [ ] \`estimate_vram()\` returns memory breakdown for any config
- [ ] Pre-training check compares estimate against available GPU memory
- [ ] Actionable suggestions when memory is insufficient
- [ ] \`--dry-run\` flag shows estimates without loading the model
- [ ] Estimates within 20% of actual usage (validate against real training runs)
- [ ] Works with all four pre-configured base models
- [ ] Unit tests for estimation formula

## Dependencies

- Depends on #8 (training pipeline)

## Estimated Effort

Medium (3-5 days). The estimation formulas are known. Calibrating them to be within 20% accuracy requires empirical testing.
